skills/repo.go

```diff
@@ -61,3 +67,102 @@ func (r *Repository) AddType(t Type) error {
 	r.types[typeName] = t
 	return nil
 }
+
+// AllSkills returns all skills of the given type.
+// If skillType is the empty string, all skills are returned.
+func (r *Repository) AllSkills(skillType string) []*Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Skill
+	for _, skillsForSnap := range r.skills {
+		for _, skill := range skillsForSnap {
+			if skillType == "" || skill.Type == skillType {
+				result = append(result, skill)
+			}
+		}
+	}
+	sort.Sort(bySkillSnapAndName(result))
+	return result
+}
+
+// Skills returns the skills offered by the named snap.
+func (r *Repository) Skills(snapName string) []*Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Skill
+	for _, skill := range r.skills[snapName] {
+		result = append(result, skill)
+	}
+	sort.Sort(bySkillSnapAndName(result))
+	return result
+}
+
+// Skill returns the specified skill from the named snap.
+func (r *Repository) Skill(snapName, skillName string) *Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.skills[snapName][skillName]
+}
+
+// AddSkill adds a skill to the repository.
+// Skill names must be valid snap names, as defined by ValidateName.
+// Skill name must be unique within a particular snap.
+func (r *Repository) AddSkill(skill *Skill) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Reject snaps with invalid names
+	if err := snap.ValidateName(skill.Snap); err != nil {
+		return err
+	}
+	// Reject skill with invalid names
+	if err := ValidateName(skill.Name); err != nil {
+		return err
+	}
+	t := r.types[skill.Type]
+	if t == nil {
+		return fmt.Errorf("cannot add skill, skill type %q is not known", skill.Type)
+	}
+	// Reject skill that don't pass type-specific sanitization
+	if err := t.Sanitize(skill); err != nil {
+		return err
+	}
+	if _, ok := r.skills[skill.Snap][skill.Name]; ok {
+		return fmt.Errorf("cannot add skill, skill name %q is in use", skill.Name)
+	}
+	if r.skills[skill.Snap] == nil {
+		r.skills[skill.Snap] = make(map[string]*Skill)
+	}
+	r.skills[skill.Snap][skill.Name] = skill
+	return nil
+}
+
+// RemoveSkill removes the named skill provided by a given snap.
+// The removed skill must exist and must not be used anywhere.
+func (r *Repository) RemoveSkill(snapName, skillName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// TODO: Ensure that the skill is not used anywhere
+	if _, ok := r.skills[snapName][skillName]; !ok {
+		return fmt.Errorf("cannot remove skill %q, no such skill", skillName)
+	}
+	delete(r.skills[snapName], skillName)
```

Comment:
`if len(r.skills[snapName]) == 0 { delete(r.skills, snapName) }`

So it doesn't grow unbounded over time..